### PR TITLE
Add ip_mask_to_prefix, ipv4_mask_to_prefix and ipv6_mask_to_prefix

### PR DESCRIPTION
**Pull request by ebarnard**

### CODE REVIEW

**Comment by achanda** [comment]
Thanks @ebarnard!
I think the API will look better if we can do something like
```
impl From<(Ipv4Addr, Ipv4Addr)> for Ipv4Network {
    fn from(pair: (Ipv4Addr, Ipv4Addr)) -> Result<Ipv4Network, IpNetworkError> {
        let mask = u32::from(pair.1);
        let prefix = (!mask).leading_zeros() as u8;
        let suffix = mask.trailing_zeros() as u8;

        if prefix + suffix != IPV4_BITS {
            Err(IpNetworkError::InvalidPrefix)
        } else {
            Ipv4Network::new(pair.0, pair.1)
        }
    }
}
```
But this does not build. What do you think? Any idea if there is a way to return error from a trait implementation?
cc @faern

**Comment by ebarnard** [comment]
I don't think it's possible to return an error from `From`.

I'm also not sure that api looks better although I agree the name `from_addr_and_mask` isn't ideal. 

It's not clear that you need to pass as (ip, mask) pair, rather than (ip, network) or (mask, ip) and as it's a trait implementation you can't document it.

File: src/ipv6.rs [hunk 1]
```diff
@@ -34,6 +35,40 @@ impl Ipv6Network {
         Self::new(addr, prefix)
     }
 
+    /// Constructs a new `Ipv4Network` from an `Ipv4Addr` and a network mask.
```
Comment (faern) [inline src/ipv6.rs]: Says Ipv4Addr..

Comment (ebarnard) [inline src/ipv6.rs]: Whoops

**Comment by faern** [comment]
You can't change the signature, so it's impossible to return a `Result` instead of the type directly. `From` does not allow failing.

I would try refactoring and making a separate `fn mask_to_prefix(mask: IpAddr) -> u8`

**Comment by achanda** [comment]
Looks like the `TryFrom` trait supports returning a `Result`, but that's not stable yet.

**Comment by faern** [comment]
@achanda And as @ebarnard pointed out, it would not allow us to put names on the two arguments, so it would be unclear what it did. But I agree the name need to be better than now.

**Comment by ebarnard** [comment]
@achanda 

I think a separate `mask_to_prefix` would be good.

It would have to be `fn mask_to_prefix(mask: IpAddr) -> Result<u8, ERR>` as an IpAddr could not be a valid prefix.

**Comment by faern** [comment]
"Could be an invalid" you mean? Yes sure. I wrote the same signature, but I posted just u8 first so you might not have gotten the update. Yes, sounds about right. It the implementation is simply a match on the type of IP and then the implementations differ then I prefer that `mask_to_prefix` is put in each struct impl and take the direct type directly. Cleaner code and probably faster to execute.

**Comment by ebarnard** [comment]
Probably worth having one with a match on IpNetwork as well as per `new`, `prefix`, `ip` etc.

**Comment by ebarnard** [comment]
I've added free functions `ip_mask_to_prefix`, `ipv4_mask_to_prefix` and `ipv6_mask_to_prefix`.

The first is for parity with `IpNetwork`

**Comment by achanda** [comment]
Thanks @ebarnard!

**Pull request merged**